### ISSUE TITLE
KVM is not enabled on Jammy Ubuntu variant

### ISSUE THREAD

**Issue by AndnixSH** [devcontainers/images#884]
When using other Ubuntu variant (Jammy) using devcontainer, provided in this repo, there is no kvm support

```
$ sudo kvm-ok
INFO: /dev/kvm does not exist
HINT:   sudo modprobe kvm_amd
/usr/sbin/kvm-ok: 93: modprobe: not found
```

but the default codespace (based on Focal) has kvm supported by default
```
$ sudo kvm-ok
INFO: /dev/kvm exists
KVM acceleration can be used
```

what should I do to enable kvm support?

**Comment by joshspicer** [comment]
Could you please provide some more information, namely:

(1) Are you using the dev containers extension, Codespaces, or some other client?
(2) What are you using `kvm-amd` for? Ciould you please explain the use-case a little bit more for our knowledge?
(3) Could you please specify the exact images/configs when you reference "other Uubntu variant" and "default codespace"? That will help us later if want to try and repro your situation?

**Comment by AndnixSH** [comment]
1) I'm using Github codespace on github.com
2) I need kvm to increase performance for QEMU for OS emulation. Without kvm, the performance is worse.
3) For jammy variant, I just used this one https://github.com/devcontainers/images/tree/main/src/base-ubuntu. For default, I create new codespace with blank template. There is no config, so it loads Ubuntu 20.04 (focal) by default

**Comment by tlaurion** [comment]
Also interested to know what difffers from the base image and why I cannot replicate the same result with kvm-ok with a 
alternative Dockerfiles:
```
# Start from the Debian "Bookworm" image
FROM mcr.microsoft.com/devcontainers/base:dev-bookworm
```

```
@tlaurion ➜ /workspaces/heads (codespace_test) $ sudo kvm
Could not access KVM kernel module: No such file or directory
kvm: failed to initialize kvm: No such file or directory
kvm: falling back to tcg
gtk initialization failed
```

**Comment by tlaurion** [comment]
@joshspicer @AndnixSH @eljog @joshaber @brettcannon @michaelcontento @Chuxel @jkeech

Any idea/pointer into what differenciate default codespace container to others which don't expose kvm properly?

My use case is able to promote codespace proper config, create proper docker image to use and customize, based on debian bookworm (which provides swtpm) in combination of kvm (which otherwise is qemu tcg which would be heavier on host) to dev/build/test Heads firmware through codespace.

Without kvm, IO even under QubesOS (which doesn't offer nested virt), even installing netinst os takes a night because qemu tcg.

TLDR: having bookworm with kvm enabled would permit to quickly test built Heads firmware under codespace. This is not directly related to name of this issue but is basically the same: only default, old, Ubuntu default docker offers kvm enabled codespace. What is missing in other docker images?

**Comment by NickBorgers** [comment]
KVM is a hypervisor provided by the Linux Kernel. Hypervisors allows the creation of isolated virtual machines which run on the same hardware. Containers are isolated environments which run on the same kernel.

I would argue that GitHub codespaces is pretty unusual for arranging itself to allow you to interact with the KVM interface. In general containers limit processes ability to interact with the kernel.

I would expect you currently have two issues with what you're trying to do:
- Your Devcontainer doesn't have access to the KVM interfaces of the kernel running it
- The KVM cli tools and supporting packages aren't installed in the Devcontainer by-default because this is a weird usecaes

Running the container as privileged might resolve the first problem, and the second would be solved by installing those packages.

This might work from somewhere 🤷 , it doesn't work on my machine. I didn't expect it to work on my machine because Docker is running in a VM spun up by Docker Desktop and it wouldn't make any sense to include the KVM kernel module b/c nested virtualization is not supported by any hypervisor I've worked with:
```
docker run --privileged --rm -it --device=/dev/kvm mcr.microsoft.com/devcontainers/base:dev-bookworm sh -c '\
apt update && \
apt -y install bridge-utils cpu-checker libvirt-clients libvirt-daemon qemu-kvm && \
kvm-ok\
'
```

Tagging a bunch of people like this is super loud and not within my experience of norms around open source code. I expect it will come off as rude.

**Comment by AndnixSH** [comment]
So can't you guys provide at least Ubuntu 22.04 machine without using Dockerfile? The default machine clearly allows you to interact witht the KVM interface, QEMU indeed works perfectly on Codespace. The only disadvantage is older Ubuntu version

**Comment by eljog** [comment]
As Nick suggested, running the container in privileged mode may help you get necessary kernel level access.
You can of course do this using `devcontainer.json` without having to customize Dockerfile.

To run container in privileged mode, add the following property to your devcontainer.json. Read more about this **[here](https://containers.dev/implementors/json_reference/)**
`"privileged": true`

To install additional packages on your container you have multiple options outside of modifying/creating a Dockerfile. The best approach would be to install them as a feature if a devcontainer feature is available.
- A list of available features can be found [here](https://containers.dev/features)
- If a feature is not already present for your use case, you can author a feature by following https://github.com/devcontainers/feature-starter.

Alternatively you may plug in your `apt-get install` commands as an `ocCreateCommand` in your devcontainer.json. Read more about lifecycle scripts [**here**](https://containers.dev/implementors/json_reference/#lifecycle-scripts)

**Comment by AndnixSH** [comment]
Thanks. `"privileged": true` is all I'm looking for, KVM is now enabled!
